src/main/java/org/kuali/kfs/krad/service/impl/DocumentServiceImpl.java

```diff
@@ -0,0 +1,1164 @@
+        }
+
+        WorkflowDocument workflowDocument = null;
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Retrieving doc id: " + documentHeaderId + " from workflow service.");
+        }
+
+        Person person = getPersonService().getPersonByPrincipalName(KRADConstants.SYSTEM_USER);
+        workflowDocument = workflowDocumentService.loadWorkflowDocument(documentHeaderId, person);
+
+        Class<? extends Document> documentClass = getDocumentClassByTypeName(workflowDocument.getDocumentTypeName());
+
+        // retrieve the Document
+        Document document = getDocumentDao().findByDocumentHeaderId(documentClass, documentHeaderId);
+
+        return postProcessDocument(documentHeaderId, workflowDocument, document);
+    }
+
+    private Class<? extends Document> getDocumentClassByTypeName(String documentTypeName) {
+        if (StringUtils.isBlank(documentTypeName)) {
+            throw new IllegalArgumentException("invalid (blank) documentTypeName");
+        }
+
+        Class<? extends Document> clazz = getDataDictionaryService().getDocumentClassByTypeName(documentTypeName);
+        if (clazz == null) {
+            throw new UnknownDocumentTypeException(
+                    "unable to get class for unknown documentTypeName '" + documentTypeName + "'");
+        }
+        return clazz;
+    }
+
+    /**
+     * Loads the Notes for the note target on this Document.
+     *
+     * @param document the document for which to load the notes
+     */
+    protected void loadNotes(Document document) {
+        if (isNoteTargetReady(document)) {
+            List<Note> notes = new ArrayList<Note>();
+            if (StringUtils.isNotBlank(document.getNoteTarget().getObjectId())) {
+                notes.addAll(getNoteService().getByRemoteObjectId(document.getNoteTarget().getObjectId()));
+            }
+            //notes created on 'disapprove' are linked to Doc Header, so this checks that even if notetype = BO
+            if (document.getNoteType().equals(NoteType.BUSINESS_OBJECT)
+                    && document.getDocumentHeader().getWorkflowDocument().isDisapproved()) {
+                notes.addAll(getNoteService().getByRemoteObjectId(document.getDocumentHeader().getObjectId()));
+            }
+
+            // KULRNE-5692 - force a refresh of the attachments
+            // they are not (non-updateable) references and don't seem to update properly upon load
+            for (Note note : notes) {
+                note.refreshReferenceObject("attachment");
+            }
+            document.setNotes(notes);
+        }
+    }
+
+    /**
+     * Performs required post-processing for every document from the documentDao
+     *
+     * @param documentHeaderId
+     * @param workflowDocument
+     * @param document
+     */
+    protected Document postProcessDocument(String documentHeaderId, WorkflowDocument workflowDocument, Document document) {
```

Comment:
Yes, that is correct. And I'm hopeful we could contribute that back to base code and eliminate this overlay, possibly as soon as next sprint.